R/predict.R

```diff
@@ -0,0 +1,152 @@
+#'
+#' The equation used for the standard error of a prediction given a row of data x is:
+#'
+#' \eqn{\sqrt(x \Sigma x')},
+#'
+#' where \eqn{\Sigma} is the estimated variance-covariance matrix from \code{lm_robust}.
+#'
+#' The prediction intervals are for a single observation at each case in newdata with error variance(s) pred.var. The the default is to assume that future observations have the same error variance as those used for fitting, which is gotten from the fit lm_robust object. If weights is supplied, the inverse of this is used as a scale factor. If the fit was weighted, the default is to assume constant prediction variance, with a warning.
+#'
+#' @examples
+#'
+#' # Set seed
+#' set.seed(42)
+#'
+#' # Simulate data
+#' n <- 10
+#' dat <- data.frame(y = rnorm(n), x = rnorm(n))
+#'
+#' # Fit lm
+#' lm_out <- lm_robust(y ~ x, data = dat)
+#' # Get predicted fits
+#' fits <- predict(lm_out, newdata = dat)
+#' # With standard errors and confidence intervals
+#' fits <- predict(lm_out, newdata = dat, se.fit = TRUE, interval = "confidence")
+#'
+#' # Use new data as well
+#' new_dat <- data.frame(x = runif(n, 5, 8))
+#' predict(lm_out, newdata = new_dat)
+#'
+#' # You can also supply custom variance weights for prediction intervals
+#' new_dat$w <- runif(n)
+#' predict(lm_out, newdata = new_dat, weights = w, interval = "prediction")
+#'
+#' @export
+predict.lm_robust <- function(
+  object,
+  newdata,
+  se.fit = FALSE,
+  interval = c("none", "confidence", "prediction"),
+  alpha = 0.05,
+  na.action = na.pass,
+  pred.var = NULL,
+  weights,
+  ...) {
+
+  # Get model matrix
+  rhs_terms <- delete.response(object$terms)
+  mf <- model.frame(rhs_terms, newdata, na.action = na.action)
+
+  # Check class of columns in newdata match those in model fit
+  if(!is.null(cl <- attr(rhs_terms, "dataClasses"))) .checkMFClasses(cl, mf)
+
+  X <- model.matrix(rhs_terms, mf, contrasts.arg = object$contrasts)
+
+  # lm_lin scaling
+  if (!is.null(object$scaled_center)) {
+    demeaned_covars <-
+      scale(
+        X[
+          ,
+          names(object$scaled_center),
+          drop = F
+          ],
+        center = object$scaled_center,
+        scale = FALSE
+      )
+
+
+    # Interacted with treatment
+    treat_name <- attr(object$terms, "term.labels")[1]
+    interacted_covars <- X[, treat_name] * demeaned_covars
+
+    X <- cbind(X[, attr(X, "assign") <= 1, drop = F],
+               demeaned_covars,
+               interacted_covars)
+  }
+
+  # Get predicted values
+  predictor <- drop(X %*% object$est)
+
+  df_resid <- object$n - object$rank
+  interval <- match.arg(interval)
+
+  if (se.fit || interval != "none") {
+
+    ret <- list()
+
+    var_fit <-
+      apply(X, 1, function(x) tcrossprod(crossprod(x, object$vcov), x))
```

Comment:
for ```x'Sx``` I usually group it the other way - the parens here are a little confusing to me, it looks like you wrote ```x'Sx'``` which seems odd.

Reply:
It's xSx', where x is a row-vector.

The reason I do `crossprod(x, object$vcov)` is that crossprod will take `x` to be a column vector when I want it to be a row vector.

Reply:
in R vectors are vectors, you usually don't need to worry about row/column vectors when writing code, but going backwards from code to math is difficult. EG:

```
> S <- matrix(1:9,3,3)
> x <- 1:3
> tcrossprod(crossprod(x, S), x)
     [,1]
[1,]  228
> x %*% S %*% x
     [,1]
[1,]  228
> x %*% crossprod(S, x) 
     [,1]
[1,]  228
```

If the intent is `xSx'` (with one transpose mark in the equation at the very end) I would suggest - 

```
>  tcrossprod( x %*% S, x)
     [,1]
[1,]  228
```

bc `x %*% tcrossprod(S,x)` throws an error.



as it's the most literal translation.